File: Find the Maximum Sum of Node Values/Find the Maximum Sum of Node Values.py
# Solution for Find the Maximum Sum of Node Values in PY

class Solution:
    def maximumValueSum(self, nums: List[int], k: int, edges: List[List[int]]) -> int:
        total_sum = 0
        count_better_with_xor = 0
        min_loss_when_forced_to_not_use_xor = sys.maxsize

        for num in nums:
            xor_value = num ^ k

            if xor_value > num:
                count_better_with_xor += 1
                total_sum += xor_value
            else:
                total_sum += num

            min_loss_when_forced_to_not_use_xor = min(min_loss_when_forced_to_not_use_xor, abs(num - xor_value))

        if count_better_with_xor % 2 == 1:
            return total_sum - min_loss_when_forced_to_not_use_xor

        return total_sum

File: Find the Maximum Sum of Node Values/Find the Maximum Sum of Node Values.java
// Solution for Find the Maximum Sum of Node Values in JAVA

class Solution {
    public long maximumValueSum(int[] nums, int k, int[][] edges) {
        long totalSum = 0;
        int countBetterWithXOR = 0;
        int minLossWhenForcedToNotUseXOR = Integer.MAX_VALUE;

        for (int num : nums) {
            int xorValue = num ^ k;

            // If applying XOR increases the value, we tentatively choose it
            if (xorValue > num) {
                countBetterWithXOR++;
                totalSum += xorValue;
            } else {
                totalSum += num;
            }

            // Track the minimum difference between original and XOR value
            minLossWhenForcedToNotUseXOR = Math.min(minLossWhenForcedToNotUseXOR, Math.abs(num - xorValue));
        }

        // If odd number of XOR improvements, one must be undone to keep parity (edges come in pairs)
        if (countBetterWithXOR % 2 == 1) {
            return totalSum - minLossWhenForcedToNotUseXOR;
        }

        return totalSum;
    }
}
